### PR TITLE
Feat(pt): Support Density fitting

### PR DESCRIPTION
**Pull request by YuzhiLiu-ai**
### Summary

This PR adds charge density prediction support to the PyTorch backend of DeePMD-kit.

The implementation enables a DPA3-based model to predict charge density values on user-defined spatial grid points. It extends the existing atom-centered energy/force framework to grid-based scalar field prediction.

The core idea is to treat each grid point as a virtual central atom. For each grid point, the model constructs a directional neighbor list from the grid point to surrounding real atoms, computes DPA3 descriptors, and applies a density fitting network to predict the charge density value at that grid location.

#### 1. New density model architecture

- Added `GridDensityModel` and `make_density_model` as model wrappers for grid-based density inference.
- Added `DensityAtomicModel` as the underlying atomic model for descriptor evaluation and density fitting.
- Grid points are treated as virtual central atoms, while real atoms serve as their neighbors.
- The model outputs charge density values on the input grid.

#### 2. Data pipeline support

- Updated `deepmd/utils/data.py` to support two additional data entries:
  - `grid`
  - `density`
- These fields bypass the standard atom-type selection logic (`type_sel`) because they are grid-based rather than atom-centered quantities.

### Example

A complete working example is provided in:

```text
examples/density/
```

#### Data format

The dataset follows the standard `deepmd/npy` format, with two additional files in each `set.000/` directory:

| File | Shape | Description |
|---|---:|---|
| `coord.npy` | `[nframes, natoms * 3]` | Atomic coordinates |
| `box.npy` | `[nframes, 9]` | Simulation cell |
| `type.raw` | `[natoms]` | Atom type indices |
| `type_map.raw` | `[ntypes]` | Type map |
| `grid.npy` | `[nframes, ngrid, 3]` | Grid point coordinates |
| `density.npy` | `[nframes, ngrid, 1]` | Charge density labels |

#### Training

```bash
cd examples/density/dpa3
dp --pt train input.json
```

Key configuration in `input.json`:

```json
{
  "model": {
    "type_map": ["Li", "Ni", "Co", "Mn", "O", "C", "H", "N", "F", "X"],
    "descriptor": {
      "type": "dpa3",
      ....
      "env_protection": 0.1
    },
    "fitting_net": {
      "type": "density",
      "neuron": [240, 240, 240]
    }
  },
  "loss": {
    "type": "grid_density",
    "start_pref_d": 1,
    "limit_pref_d": 1
  }
}
```

Notes:

- The last element in `type_map`, `"X"`, is reserved for virtual grid points.
- During inference, grid points are assigned `atype = ntypes - 1`, so that the descriptor treats them as distinct from real atoms.
- For DPA3, `env_protection` is set to `0.1` to avoid numerical instability when a grid point is very close to, or coincides with, a real atom position.

#### Evaluation

```bash
cd examples/density
python dptest_density_script.py \
    dpa3/model.ckpt-100.pt \
    dataset/qm9/C7H15NO_val \
    --ratio 0.1 \
    --output val_result.txt
```

### Notes for reviewers

This implementation is optional and does not change the standard energy/force training workflow. The density model is activated only when the corresponding density model and loss types are specified in the input configuration.

## Summary by CodeRabbit

* **New Features**
  * Added grid-based density prediction and training support, including model, fitting, loss, data loading, and evaluation updates.
  * Added an example workflow with ready-to-use configs, sample data, and a density evaluation script.

* **Bug Fixes / Improvements**
  * Increased distributed training startup timeout for more reliable multi-process launches.
  * Improved handling of grid/density inputs throughout loading and inference paths.
  * Reduced console noise by suppressing automatic dataset summary printing.

* **Documentation**
  * Added a full density example guide with setup, training, evaluation, and export steps.

### CODE REVIEW

**Comment by njzjz** [comment]
Possible reviewers based on changed lines, exact file history, and exact-file review history:

- @iProzd — 101 commits on changed files (`deepmd/infer/deep_pot.py`, `deepmd/pt/entrypoints/main.py`, `deepmd/pt/infer/deep_eval.py`, `deepmd/pt/loss/__init__.py`, `deepmd/pt/model/model/__init__.py`, `deepmd/pt/model/task/__init__.py`, `deepmd/pt/train/training.py`, `deepmd/pt/train/wrapper.py`, `deepmd/pt/utils/dataloader.py`, `deepmd/pt/utils/stat.py`, `deepmd/utils/argcheck.py`, `deepmd/utils/data.py`).
- @wanghan-iapcm — 30 commits on changed files; 47 reviews on exact changed files (`deepmd/infer/deep_pot.py`, `deepmd/pt/entrypoints/main.py`, `deepmd/pt/infer/deep_eval.py`, `deepmd/pt/model/atomic_model/__init__.py`, `deepmd/pt/model/model/__init__.py`, `deepmd/pt/model/task/__init__.py`, `deepmd/pt/train/training.py`, `deepmd/pt/train/wrapper.py`, `deepmd/pt/utils/stat.py`, `deepmd/utils/argcheck.py`, `deepmd/utils/data.py`).

No review request was made automatically.

Coding agent: Codex
Codex version: codex-cli 0.144.4
Model: gpt-5.6-sol
Reasoning effort: xhigh
